### DISCUSSION TITLE
Getting captured `Path` from extensions does not work

### DISCUSSION THREAD

**Discussion by kckeiks** [tokio-rs/axum#2032] [Q&A] [answered]
### Version

```
axum v0.6.18
axum-core v0.3.4
```
### Platform

MacOS 13.0 

### Description

Getting capture path from extension does not work but getting MatchedPath does work. Is this expected? If so, would be great if this worked for `Service` implementations. Any workaround?

```rust
#[derive(Clone)]
struct TestService;

impl Service<Request<Body>> for TestService {
    type Response = Response;
    type Error = Infallible;
    type Future = Pin<Box<dyn Future<Output = Result<Response, Infallible>> + Send>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: Request<Body>) -> Self::Future {
        let matched_path = req.extensions().get::<MatchedPath>().unwrap();
        // let path = req.extensions().get::<Path<String>>().unwrap(); // returns None
        let msg = format!("Result: {:?}", matched_path.as_str());
        Box::pin(async move { Ok(msg.into_response())})
    }
}

#[tokio::main]
async fn main() {
    // build our application with a single route
    let app = Router::new().route_service("/:id", TestService);

    // run it with hyper on localhost:3000
    axum::Server::bind(&"0.0.0.0:8081".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```

**Comment by kckeiks** [comment]
It seems to work with`extract_parts_with_state` but this returns a future.

**Comment by jplatte** [comment]
Are you looking for the whole request path? That would be `req.uri().path()`. `Path` is for extracting path parameters, and it's not possible for axum to put it in the request extensions since it has a generic parameter chosen by the user independently of the routing.

**Comment by kckeiks** [comment]
> Are you looking for the whole request path? That would be `req.uri().path()`. `Path` is for extracting path parameters, and it's not possible for axum to put it in the request extensions since it has a generic parameter chosen by the user independently of the routing.

Not the whole request path but the captured value in the path. In the example above, I'd like the value of `:id`.

**Comment by jplatte** [comment]
Right, there is no way to do that without async at the moment. I think it would be easy enough to make it possible, but not sure if it's really worth it.

In your example above, why do you want to extract the path outside the `async move` block?

**Comment by kckeiks** [comment]
That's just an example. I'm my application I'm using a Future that will be polled.

**Comment by davidpdrsn** [comment, marked as answer]
> Getting capture path from extension does not work but getting MatchedPath does work. Is this expected?

Yep that is expected. `Path` needs to do more work than `MatchedPath` and we only wanna do that when you actually extract the path, rather than always do it up front.

So you need to run the future from `Path::from_request_parts` as part of generating the response. I'd recommend using [`axum::middleware::from_fn`](https://docs.rs/axum/latest/axum/middleware/fn.from_fn.html) which makes it easy to run extractors.
